File: services/api/src/main/java/com/voltpilot/api/web/Speicherschonung.java
package com.voltpilot.api.web;

import java.math.BigDecimal;
import java.util.Map;

/**
 * The customer-facing Speicherschonung presets (FK4): "Umgang mit dem Speicher"
 * as a ladder over the ONE existing optimizer parameter
 * {@code asset.wear_cost_ct_per_kwh} (ct per kWh cycled; the optimizer reads it
 * per 15-min cycle, so a preset takes effect on the next plan with NO optimizer
 * change). The ladder is derived from the tipping-point analysis in the
 * vp-solver-xlsx-f2 audit (§4.4): up to ~10 ct the daily PV-shift dispatch is
 * invariant, trimming starts around ~12 ct - so the presets differentiate only
 * on MARGINAL opportunities (flat spreads, micro-arbitrage), which is exactly
 * where "schonend vs. aggressiv" belongs.
 *
 * <p>This is THE mapping constant - both directions (preset -> ct on write,
 * stored ct -> preset on read) derive from {@link #PRESET_WEAR_CT}. The read
 * side maps NULL to {@code ausgewogen} (the stored NULL means "platform
 * default", which is the same 4.0 ct), and any other value to
 * {@link #INDIVIDUELL} - an admin set a custom value via the optimizer-config
 * surface; the portal shows it honestly and a customer preset pick simply
 * overwrites it. The raw ct value itself stays admin-only (optimizer-config).
 */
public final class Speicherschonung {

    /** The preset ladder: preset name -> wear cost in ct per kWh cycled. */
    public static final Map<String, BigDecimal> PRESET_WEAR_CT = Map.of(
            "aggressiv", new BigDecimal("1"),
            "ausgewogen", new BigDecimal("4"),
            "schonend", new BigDecimal("8"));

    /** NULL stored wear = the platform default 4.0 ct = the ausgewogen preset. */
    public static final String DEFAULT_PRESET = "ausgewogen";

    /** A stored value outside the ladder: an admin-configured custom wear cost. */
    public static final String INDIVIDUELL = "individuell";

    private Speicherschonung() {
    }

    /** The wear cost a preset writes; the caller validates the preset name. */
    public static BigDecimal wearCtFor(String preset) {
        BigDecimal ct = PRESET_WEAR_CT.get(preset);
        if (ct == null) {
            throw new IllegalArgumentException("Unknown Speicherschonung preset: " + preset);
        }
        return ct;
    }

    /** The effective preset for a stored wear cost (numeric match, so 4.000 = ausgewogen). */
    public static String presetFor(BigDecimal storedWearCt) {
        if (storedWearCt == null) {
            return DEFAULT_PRESET;
        }
        return PRESET_WEAR_CT.entrySet().stream()
                .filter(e -> e.getValue().compareTo(storedWearCt) == 0)
                .map(Map.Entry::getKey)
                .findFirst()
                .orElse(INDIVIDUELL);
    }
}

File: frontend/portal/src/speicherschonung.ts
/**
 * Speicherschonung (FK4): the customer's "Umgang mit dem Speicher" choice,
 * offered as three presets the backend maps onto the optimizer's battery wear
 * cost. The consequence sentences are honest derivations from the solver audit
 * (vp-solver-xlsx-f2 §4.4): on typical PV days the daily dispatch is invariant
 * across the ladder, so the presets differentiate only on marginal
 * opportunities (flat spreads, micro-arbitrage). Customer copy deliberately
 * carries NO ct/kWh numbers - the raw value stays admin-only (optimizer-config).
 *
 * The EFFECTIVE preset is derived server-side (`SiteAsset.speicherschonung`):
 * one of the three preset names, or 'individuell' when an admin configured a
 * custom wear cost - shown honestly, and a customer preset pick overwrites it.
 */

export type SpeicherschonungPreset = 'aggressiv' | 'ausgewogen' | 'schonend';

export interface SpeicherschonungOption {
  value: SpeicherschonungPreset;
  label: string;
  recommended: boolean;
  /** One honest German consequence sentence (no numbers, no jargon). */
  sentence: string;
}

export const SPEICHERSCHONUNG_OPTIONS: SpeicherschonungOption[] = [
  {
    value: 'aggressiv',
    label: 'Aggressiv',
    recommended: false,
    sentence:
      'Jede Gelegenheit wird genutzt; der Speicher arbeitet am meisten und altert am schnellsten.',
  },
  {
    value: 'ausgewogen',
    label: 'Ausgewogen',
    recommended: true,
    sentence:
      'Tägliche Solar-Verschiebung und lohnende Gelegenheiten; unwirtschaftliche Mini-Zyklen ' +
      'unterbleiben. Auf typischen PV-Anlagen entgeht gegenüber „Aggressiv“ praktisch kein Ertrag.',
  },
  {
    value: 'schonend',
    label: 'Schonend',
    recommended: false,
    sentence: 'Der Speicher arbeitet nur, wenn es sich deutlich lohnt – längste Lebensdauer.',
  },
];

/** The label an admin-configured custom wear cost reads as. */
export const SPEICHERSCHONUNG_INDIVIDUELL_LABEL = 'Individuell (durch VoltPilot konfiguriert)';

/**
 * Read-view label for the server-derived effective preset. Null/unknown falls
 * back to Ausgewogen (the server derives a stored NULL the same way).
 */
export function speicherschonungLabel(value: string | null | undefined): string {
  if (value === 'individuell') return SPEICHERSCHONUNG_INDIVIDUELL_LABEL;
  const opt =
    SPEICHERSCHONUNG_OPTIONS.find((o) => o.value === value) ??
    SPEICHERSCHONUNG_OPTIONS.find((o) => o.value === 'ausgewogen')!;
  return opt.recommended ? `${opt.label} (empfohlen)` : opt.label;
}

/**
 * The preset a radio group can pre-select; null for 'individuell' (nothing
 * pre-selected - the presets are offered, picking one overwrites the custom
 * value) and for null/unknown values.
 */
export function presetOf(value: string | null | undefined): SpeicherschonungPreset | null {
  return SPEICHERSCHONUNG_OPTIONS.some((o) => o.value === value)
    ? (value as SpeicherschonungPreset)
    : null;
}
